From 8aa8aeb6fa9b40058598b5f08abc338ba6f29538 Mon Sep 17 00:00:00 2001
From: oli <olit@google.com>
Date: Wed, 21 May 2025 14:38:32 +0000
Subject: [PATCH] Add resolveActivityAsUserForExplicitType api to pm

allow checking for resolved activity for an intent with a specific type

allowing the intent to declare its own type can lead to a security
vulnerability if the intent changes its type after the IntentForwarder#canForward
check.

Use the new API in IntentForwarderActivity to prevent checking the intent's type twice.

Update other call sites of IntentForwarder#canForward to extract the resolved type from the intent and pass in to to the method call

Check the launch package rather than calling package

sanitize intent selector as well as received intent

Bug: 403565650 407763772 397216638
Flag: EXEMPT bugfix
Test: manually tested
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:57233409d9bb5aef50373760a9b6551aa618c808)
Merged-In: I8453f25ed5ddd938718f4d2be4983c881778f281
Change-Id: I8453f25ed5ddd938718f4d2be4983c881778f281
---
 .../app/ApplicationPackageManager.java        | 28 +++++++++++----
 .../android/content/pm/PackageManager.java    | 32 +++++++++++++++--
 .../internal/app/IntentForwarderActivity.java | 34 ++++++++++++++-----
 .../app/IntentForwarderActivityTest.java      | 20 +++++++++--
 4 files changed, 95 insertions(+), 19 deletions(-)

diff --git a/core/java/android/app/ApplicationPackageManager.java b/core/java/android/app/ApplicationPackageManager.java
index d641a3b469a6..3d4789f8a53e 100644
--- a/core/java/android/app/ApplicationPackageManager.java
+++ b/core/java/android/app/ApplicationPackageManager.java
@@ -1386,22 +1386,36 @@ public class ApplicationPackageManager extends PackageManager {
 
     @Override
     public ResolveInfo resolveActivity(Intent intent, ResolveInfoFlags flags) {
-        return resolveActivityAsUser(intent, flags, getUserId());
+        return resolveActivityAsUser(intent, /* resolvedType= */ null, flags, getUserId());
     }
 
     @Override
     public ResolveInfo resolveActivityAsUser(Intent intent, int flags, int userId) {
-        return resolveActivityAsUser(intent, ResolveInfoFlags.of(flags), userId);
+        return resolveActivityAsUser(intent, /* resolvedType= */ null, ResolveInfoFlags.of(flags),
+                userId);
     }
 
     @Override
     public ResolveInfo resolveActivityAsUser(Intent intent, ResolveInfoFlags flags, int userId) {
+        return resolveActivityAsUser(intent, /* resolvedType= */ null, flags, userId);
+    }
+
+    @Override
+    public ResolveInfo resolveActivityAsUser(Intent intent, String resolvedType,
+            int flags, int userId) {
+        return resolveActivityAsUser(intent, resolvedType,
+                ResolveInfoFlags.of(flags), userId);
+    }
+
+    @Override
+    public ResolveInfo resolveActivityAsUser(Intent intent, String resolvedType,
+            ResolveInfoFlags flags, int userId) {
         try {
-            return mPM.resolveIntent(
-                intent,
-                intent.resolveTypeIfNeeded(mContext.getContentResolver()),
-                updateFlagsForComponent(flags.getValue(), userId, intent),
-                userId);
+            return mPM.resolveIntent(intent,
+                    resolvedType == null
+                        ? intent.resolveTypeIfNeeded(mContext.getContentResolver())
+                        : resolvedType,
+                    updateFlagsForComponent(flags.getValue(), userId, intent), userId);
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
         }
diff --git a/core/java/android/content/pm/PackageManager.java b/core/java/android/content/pm/PackageManager.java
index f224f8a189f1..15e412a0148b 100644
--- a/core/java/android/content/pm/PackageManager.java
+++ b/core/java/android/content/pm/PackageManager.java
@@ -6861,12 +6861,20 @@ public abstract class PackageManager {
      * Intent.resolveActivity(PackageManager)} do.
      * </p>
      *
+     * Use {@link #resolveActivityAsUser(Intent, String, ResolveInfoFlags, int)}
+     * when long flags are needed.
+     *
      * @param intent An intent containing all of the desired specification
      *            (action, data, type, category, and/or component).
+     * @param resolvedType A nullable resolved type for the intent's data.
+     *            Specified explicitly when the data type contained in the
+     *            intent cannot be trusted. If null is provided, the type will
+     *            be obtained from the intent using
+     *            {@link Intent#resolveTypeIfNeeded}.
      * @param flags Additional option flags to modify the data returned. The
      *            most important is {@link #MATCH_DEFAULT_ONLY}, to limit the
      *            resolution to only those activities that support the
-     *            {@link android.content.Intent#CATEGORY_DEFAULT}.
+     *            {@link Intent#CATEGORY_DEFAULT}.
      * @param userId The user id.
      * @return Returns a ResolveInfo object containing the final activity intent
      *         that was determined to be the best action. Returns null if no
@@ -6876,6 +6884,26 @@ public abstract class PackageManager {
      * @hide
      * @deprecated Use {@link #resolveActivityAsUser(Intent, ResolveInfoFlags, int)} instead.
      */
+    @Nullable
+    public ResolveInfo resolveActivityAsUser(@NonNull Intent intent,
+            @Nullable String resolvedType, int flags, @UserIdInt int userId) {
+        return resolveActivityAsUser(intent, resolvedType, ResolveInfoFlags.of(flags), userId);
+    }
+
+    /**
+     * See {@link #resolveActivityAsUser(Intent, String, int, int)}.
+     * @hide
+     */
+    @Nullable
+    public ResolveInfo resolveActivityAsUser(@NonNull Intent intent,
+            @Nullable String resolvedType, @NonNull ResolveInfoFlags flags, @UserIdInt int userId) {
+        throw new UnsupportedOperationException(
+                "resolveActivityAsUser not implemented in subclass");
+    }
+     /**
+     * See {@link #resolveActivityAsUser(Intent, String, int, int)}.
+     * @hide
+     */
     @Deprecated
     @SuppressWarnings("HiddenAbstractMethod")
     @Nullable
@@ -6884,7 +6912,7 @@ public abstract class PackageManager {
             int flags, @UserIdInt int userId);
 
     /**
-     * See {@link #resolveActivityAsUser(Intent, int, int)}.
+     * See {@link #resolveActivityAsUser(Intent, String, int, int)}.
      * @hide
      */
     @Nullable
diff --git a/core/java/com/android/internal/app/IntentForwarderActivity.java b/core/java/com/android/internal/app/IntentForwarderActivity.java
index 1ecd47e2b81b..afe2d83a9ec8 100644
--- a/core/java/com/android/internal/app/IntentForwarderActivity.java
+++ b/core/java/com/android/internal/app/IntentForwarderActivity.java
@@ -30,7 +30,6 @@ import android.app.AppGlobals;
 import android.app.admin.DevicePolicyManager;
 import android.compat.annotation.UnsupportedAppUsage;
 import android.content.ComponentName;
-import android.content.ContentResolver;
 import android.content.Intent;
 import android.content.pm.ActivityInfo;
 import android.content.pm.IPackageManager;
@@ -133,8 +132,9 @@ public class IntentForwarderActivity extends Activity  {
         }
 
         final int callingUserId = getUserId();
+        String resolvedType = intentReceived.resolveTypeIfNeeded(getContentResolver());
         final Intent newIntent = canForward(intentReceived, getUserId(), targetUserId,
-                mInjector.getIPackageManager(), getContentResolver());
+                mInjector.getIPackageManager(), resolvedType);
 
         if (newIntent == null) {
             Slog.wtf(TAG, "the intent: " + intentReceived + " cannot be forwarded from user "
@@ -145,7 +145,11 @@ public class IntentForwarderActivity extends Activity  {
 
         newIntent.prepareToLeaveUser(callingUserId);
         final CompletableFuture<ResolveInfo> targetResolveInfoFuture =
-                mInjector.resolveActivityAsUser(newIntent, MATCH_DEFAULT_ONLY, targetUserId);
+                mInjector.resolveActivityAsUser(
+                        newIntent,
+                        resolvedType,
+                        MATCH_DEFAULT_ONLY,
+                        targetUserId);
         targetResolveInfoFuture
                 .thenApplyAsync(targetResolveInfo -> {
                     if (isResolverActivityResolveInfo(targetResolveInfo)) {
@@ -248,6 +252,9 @@ public class IntentForwarderActivity extends Activity  {
                 ? targetUserId : callingUserId;
         int selectedProfile = findSelectedProfile(className);
         sanitizeIntent(intentReceived);
+        if (intentReceived.getSelector() != null) {
+            sanitizeIntent(intentReceived.getSelector());
+        }
         intentReceived.putExtra(EXTRA_SELECTED_PROFILE, selectedProfile);
         intentReceived.putExtra(EXTRA_CALLING_USER, UserHandle.of(callingUserId));
         startActivityAsCaller(intentReceived, null, false, userId);
@@ -313,20 +320,20 @@ public class IntentForwarderActivity extends Activity  {
      * forwarding if it can be forwarded, {@code null} otherwise.
      */
     static Intent canForward(Intent incomingIntent, int sourceUserId, int targetUserId,
-            IPackageManager packageManager, ContentResolver contentResolver)  {
+            IPackageManager packageManager, String resolvedType)  {
         Intent forwardIntent = new Intent(incomingIntent);
         forwardIntent.addFlags(
                 Intent.FLAG_ACTIVITY_FORWARD_RESULT | Intent.FLAG_ACTIVITY_PREVIOUS_IS_TOP);
         sanitizeIntent(forwardIntent);
 
         if (!canForwardInner(forwardIntent, sourceUserId, targetUserId, packageManager,
-                contentResolver)) {
+                resolvedType)) {
             return null;
         }
         if (forwardIntent.getSelector() != null) {
             sanitizeIntent(forwardIntent.getSelector());
             if (!canForwardInner(forwardIntent.getSelector(), sourceUserId, targetUserId,
-                    packageManager, contentResolver)) {
+                    packageManager, resolvedType)) {
                 return null;
             }
         }
@@ -334,11 +341,10 @@ public class IntentForwarderActivity extends Activity  {
     }
 
     private static boolean canForwardInner(Intent intent, int sourceUserId, int targetUserId,
-            IPackageManager packageManager, ContentResolver contentResolver) {
+            IPackageManager packageManager, String resolvedType) {
         if (Intent.ACTION_CHOOSER.equals(intent.getAction())) {
             return false;
         }
-        String resolvedType = intent.resolveTypeIfNeeded(contentResolver);
         try {
             if (packageManager.canForwardTo(
                     intent, resolvedType, sourceUserId, targetUserId)) {
@@ -419,6 +425,15 @@ public class IntentForwarderActivity extends Activity  {
             return IntentForwarderActivity.this.getPackageManager();
         }
 
+        @Override
+        @Nullable
+        public CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent,
+                String resolvedType, int flags, int userId) {
+            return CompletableFuture.supplyAsync(
+                    () -> getPackageManager().resolveActivityAsUser(intent,
+                            resolvedType, flags, userId));
+        }
+
         @Override
         @Nullable
         public CompletableFuture<ResolveInfo> resolveActivityAsUser(
@@ -440,6 +455,9 @@ public class IntentForwarderActivity extends Activity  {
 
         PackageManager getPackageManager();
 
+        CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent,
+                String resolvedType, int flags, int userId);
+
         CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent, int flags, int userId);
 
         void showToast(String message, int duration);
diff --git a/core/tests/coretests/src/com/android/internal/app/IntentForwarderActivityTest.java b/core/tests/coretests/src/com/android/internal/app/IntentForwarderActivityTest.java
index a663095aa893..37f5b3a12031 100644
--- a/core/tests/coretests/src/com/android/internal/app/IntentForwarderActivityTest.java
+++ b/core/tests/coretests/src/com/android/internal/app/IntentForwarderActivityTest.java
@@ -267,9 +267,19 @@ public class IntentForwarderActivityTest {
         IntentForwarderWrapperActivity activity = mActivityRule.launchActivity(intent);
 
         ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
-        verify(mIPm).canForwardTo(
+        InstrumentationRegistry.getInstrumentation().waitForIdleSync();
+        verify(mIPm, times(2)).canForwardTo(
                 intentCaptor.capture(), nullable(String.class), anyInt(), anyInt());
-        assertEquals(Intent.ACTION_VIEW, intentCaptor.getValue().getAction());
+        List<Intent> capturedIntents = intentCaptor.getAllValues();
+        // Verify root intent is checked and sanitized
+        assertEquals(Intent.ACTION_MAIN, capturedIntents.get(0).getAction());
+        assertNull(capturedIntents.get(0));
+        assertNull(capturedIntents.get(0).getPackage());
+        // Verify selector is checked and sanitized
+        assertEquals(Intent.ACTION_VIEW, capturedIntents.get(1).getAction());
+        assertNull(capturedIntents.get(1));
+        assertNull(capturedIntents.get(1).getPackage());
+        InstrumentationRegistry.getInstrumentation().waitForIdleSync();
 
         assertNotNull(activity.mStartActivityIntent);
         assertEquals(Intent.ACTION_MAIN, activity.mStartActivityIntent.getAction());
@@ -676,6 +686,12 @@ public class IntentForwarderActivityTest {
             return mPm;
         }
 
+        @Override
+        public CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent,
+                String resolvedType, int flags, int userId) {
+            return resolveActivityAsUser(intent, flags, userId);
+        }
+
         @Override
         public CompletableFuture<ResolveInfo> resolveActivityAsUser(
                 Intent intent, int flags, int userId) {
-- 
2.34.1

